torchmultimodal/modules/layers/quantisation.py

```diff
@@ -0,0 +1,61 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import nn
+
+
+class Quantisation(nn.Module):
+    """
+    Embedding layer that takes in a collection of flattened vectors and finds closest embedding vectors
+    to each flattened vector and outputs those selected embedding vectors. Also known as vector quantisation.
+    """
+
+    def __init__(self, num_embeddings, embedding_dim):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.num_embeddings = num_embeddings
+
+        self.embedding = nn.Embedding(self.num_embeddings, self.embedding_dim)
+        self.embedding.weight.data.normal_()
+        self.quantised_vectors = None
+
+    def forward(self, x_flat):
+        x_shape = x_flat.shape
+        # channel dimension should be embedding dim so that each element in encoder
+        # output volume gets associated with single embedding vector
+        assert (
+            x_shape[-1] == self.embedding_dim
+        ), f"Expected {x_shape[-1]} to be embedding size of {self.embedding_dim}"
+
+        # Calculate distances from each encoder output vector to each embedding vector, ||x - emb||^2
+        w_t = self.embedding.weight.t()
+        distances = (
+            torch.sum(x_flat ** 2, dim=1, keepdim=True)
+            - 2 * torch.matmul(x_flat, w_t)
+            + torch.sum(w_t ** 2, dim=0, keepdim=True)
+        )
+
+        # Encoding - select closest embedding vectors
+        encoding_indices = torch.argmin(distances, dim=1).unsqueeze(1)
```

Comment:
Not sure if I understand the steps below this line. I suppose here we want to:
- find the embeddings that are closest to the input tensors (aka "quantization")
- find the corresponding indices (may not seem obvious why we need this but a downstream module may use them for other purposes)

In addition, the embeddings need to be "moved" towards the encoder outputs (similar to KL divergence loss in Kingma's VAE paper in the continuous space). I think OpenAI's [implementation](https://github.com/openai/jukebox/blob/08efbbc1d4ed1a3cef96e08a931944c8b4d63bb3/jukebox/vqvae/bottleneck.py#L173) is a bit hand-waving that omits the burden to actually compute this loss but instead merely "update" the embeddings by the delta with the encoder output. But it's ok to stick with openAI for this exercise. And if we add this logic, please also include in the docstring description.